### PR TITLE
[#151572737] Test Environment Setup in Yojimbo

### PR DESCRIPTION
**Pull request by cchung89**
Setting up the Skillz version test environment for Game Sync. Changed all the unit test, server, and client files to use the Skillz messages instead. Modified the object type for message to Skillz messages in files that use the original test messages. Modified the `premake5.lua` to make build project with our messages header file `message.h`.
Run all the projects with the new setup with premake5 commands and they all are successfully built. Run cppcheck and scan-build to make sure no extra issues occur. Run the unit tests and they all pass with the Skillz messages. Run soak and messages are successfully send from the server to client and from client to server. Run the server and client and see they are successfully start up and the client is able to connect to the server without problem. Multiple clients are able to connect to the the server at once successfully. The valgrind shows no leaks. All the premake5 commands can be run successfully with the new changes.

@mpope9 First Round
@kkaminski Second Round

### CODE REVIEW

File: client.cpp [hunk 1]
```diff
@@ -91,6 +91,36 @@ int ClientMain( int argc, char * argv[] )
      
         time += deltaTime;
 
+        Message * message = client.ReceiveMessage(0);
+
+        if (message)
```
Comment (mpope9) [inline client.cpp]: Nit spaces around parens.

File: client.cpp [hunk 2]
```diff
@@ -91,6 +91,36 @@ int ClientMain( int argc, char * argv[] )
...
+        {
+            switch (message->GetType())
```
Comment (mpope9) [inline client.cpp]: Nit spaces around parens

File: client.cpp [hunk 3]
```diff
@@ -91,6 +91,36 @@ int ClientMain( int argc, char * argv[] )
...
+            {
+                case SKILLZ_MESSAGE:
+                {
+                    SkillzMessage * skillzMessage = (SkillzMessage*) message;
+                    printf("Received skillz message # %d\n", skillzMessage->sequence );
```
Comment (mpope9) [inline client.cpp]: Nit spaces around parens

File: client.cpp [hunk 4]
```diff
@@ -91,6 +91,36 @@ int ClientMain( int argc, char * argv[] )
...
+                }
+                break;
+
+                case GAME_START:
+                {
+                    GameStart * gameStartMessage = (GameStart*) message;
+                    printf("Game Start match # %d from server\n", gameStartMessage->matchId );
```
Comment (mpope9) [inline client.cpp]: Nit spaces around parens.

File: client.cpp [hunk 5]
```diff
@@ -91,6 +91,36 @@ int ClientMain( int argc, char * argv[] )
...
+                }
+                break;
+
+                case SKILLZ_BLOCK_MESSAGE:
+                {
+                    SkillzBlockMessage * skillzBlockMessage = (SkillzBlockMessage*) message;
+                    printf("Received skillz block message # %d\n", skillzBlockMessage->sequence );
```
Comment (mpope9) [inline client.cpp]: Nit spaces around parens

File: messages.h [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+
+    /**
+     * TODO: define what should be put into the Game Start message
+     * The Game Start message should include the matchId of the server
+     * and the two clientIds to make sure the server sent the "Game Start"
+     * message to the right players
+     * Currently, the matchId is hard-code for testing usage, but will be
+     * changed to use the actual matchId of the server
+     */
+    GameStart()
```
Comment (mpope9) [inline messages.h]: Nit use initialization list for constructor.

File: messages.h [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+    }
+
+    YOJIMBO_VIRTUAL_SERIALIZE_FUNCTIONS();
+};
+
+struct SkillzBlockMessage : public BlockMessage
+{
+    uint16_t sequence;
+
+    SkillzBlockMessage()
```
Comment (mpope9) [inline messages.h]: Nit use initialization list

File: messages.h [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+    const int modulus = sizeof( messageBitsArray ) / sizeof( int );
+    const int index = sequence % modulus;
+    return messageBitsArray[index];
+}
+
+struct SkillzMessage : public Message
+{
+    uint16_t sequence;
+
+    SkillzMessage()
```
Comment (mpope9) [inline messages.h]: Nit use initialization list

File: server.cpp [hunk 9]
```diff
@@ -68,6 +68,38 @@ int ServerMain()
 
         time += deltaTime;
 
+        // Do stuff here ====================================
+        const int numClients = server.GetNumConnectedClients();
+
+        for (int i = 0; i < numClients; ++i)
+        {
+            Message * message = server.ReceiveMessage(i, 0);
```
Comment (mpope9) [inline server.cpp]: Nit spaces around parens

File: server.cpp [hunk 10]
```diff
@@ -68,6 +68,38 @@ int ServerMain()
...
+
+            if ( !message )
+                continue;
+
+            switch (message->GetType())
```
Comment (mpope9) [inline server.cpp]: Nit spaces around parens

File: soak.cpp [hunk 11]
```diff
@@ -113,7 +113,7 @@ int SoakMain()
 
 					if ( rand() % 25 )
 					{
-						TestMessage * message = (TestMessage*) client.CreateMessage( TEST_MESSAGE );
+                        SkillzMessage * message = (SkillzMessage*) client.CreateMessage( SKILLZ_MESSAGE );
```
Comment (mpope9) [inline soak.cpp]: Nit tab

File: soak.cpp [hunk 12]
```diff
@@ -123,7 +123,7 @@ int SoakMain()
 					}
 					else
 					{
-						TestBlockMessage * blockMessage = (TestBlockMessage*) client.CreateMessage( TEST_BLOCK_MESSAGE );
+                        SkillzBlockMessage * blockMessage = (SkillzBlockMessage*) client.CreateMessage( SKILLZ_BLOCK_MESSAGE );
```
Comment (mpope9) [inline soak.cpp]: Nit tab

File: yojimbo.h [hunk 13]
```diff
@@ -158,7 +158,7 @@ struct reliable_endpoint_t;
 
 namespace yojimbo
 {
-    const int MaxClients = 64;                                      ///< The maximum number of clients supported by this library. You can increase this if you want, but this library is designed around patterns that work best for [2,64] player games. If your game has less than 64 clients, reducing this will save memory.
+    const int MaxClients = 64;                                       ///< The maximum number of clients supported by this library. You can increase this if you want, but this library is designed around patterns that work best for [2,64] player games. If your game has less than 64 clients, reducing this will save memory.
```
Comment (mpope9) [inline yojimbo.h]: Nit space

**Review by mpope9 — COMMENTED** [review]
Is shared.h used anywhere still?

**Comment by cchung89** [comment]
I am keeping `shared.h` for now since I am not sure if I should delete any existing file. But since it is mostly the same, I will delete it and see if that will affect anything at all.

**Comment by mpope9** [comment]
Okay.

**Comment by kkaminski** [comment]
Isn't `messages.h` a copy of `shared.h` with all the examples stripped out?  I think we should just modify `shared.h` to have the `messages.h` content.

File: messages.h [hunk 14]
```diff
@@ -0,0 +1,168 @@
...
+    uint16_t matchId;
+
+    /**
+     * TODO: define what should be put into the Game Start message
+     * The Game Start message should include the matchId of the server
+     * and the two clientIds to make sure the server sent the "Game Start"
+     * message to the right players
+     * Currently, the matchId is hard-code for testing usage, but will be
+     * changed to use the actual matchId of the server
+     */
```
Comment (kkaminski) [inline messages.h]: I think you did define this in your doc and as far as we know it only needs the match ID.  Therefore, please remove this comment block

Comment (kkaminski) [reply messages.h]: Besides the bit about the hard-coded match ID

**Comment by cchung89** [comment]
I delete the messages.h and use the original shared.h with the Skillz messages added instead. Build all the files and check all of them work as expected. The tests are passing successfully.

**Comment by mpope9** [comment]
What does cppcheck and valgrind say?

**Comment by mpope9** [comment]
Nvm I see your valgrind comment

**Comment by cchung89** [comment]
I did another check with them, they are good.

File: soak.cpp [hunk 15]
```diff
@@ -112,8 +112,8 @@ int SoakMain()
 						break;
 
 					if ( rand() % 25 )
-					{
-						TestMessage * message = (TestMessage*) client.CreateMessage( TEST_MESSAGE );
+                    {
```
Comment (mpope9) [inline soak.cpp]: Nit tab

**Comment by mpope9** [comment]
🐘

File: shared.h [hunk 16]
```diff
@@ -75,18 +72,31 @@ struct TestMessage : public Message
...
 
+    YOJIMBO_VIRTUAL_SERIALIZE_FUNCTIONS();
+};
+
+// Game Start Message sent by the server to the clients
+struct GameStart : public BlockMessage
+{
+    uint16_t matchId;
+
+    GameStart() : matchId( 12345 ) {}
```
Comment (kkaminski) [inline shared.h]: Mentioned before but we need a TODO for this hard-coded match ID.

**Comment by kkaminski** [comment]
Do we need a separate story to clean up the cruft in `game-sync-service`?

**Comment by cchung89** [comment]
The only things I think are just taking out the unnecessary files in the game-sync-service like `server.cpp`, `soak.cpp`, etc. since the changes occur directly in the Yojimbo now, but that still will need a different PR though since it is in a different repo.

**Comment by kkaminski** [comment]
@cchung89 what's going on with https://github.com/skillz/game-sync-service/pull/7 then?

**Comment by cchung89** [comment]
I was going to close that since the changes should be in the Yojimbo now, but I think that can be used for clean up in the game-sync-service instead.

**Comment by kkaminski** [comment]
OK I just wanted resolution on that cleanup: 👍

**Pull request merged**